### ISSUE TITLE
Memory leak in Task.Run

### ISSUE THREAD

**Issue by omfgicbf** [mono/mono#16042]
## Steps to Reproduce

1. Run the code below.
2. Watch memory consumption over an extended period.

```C#
static void Main(string[] args)
{
    while (true)
    {
        Task.Run(() =>
        {
            Thread.Sleep(50);
        });

        Thread.Sleep(40);
    }
}
```

Possibly related to an issue discussed in the notes of #9631, where @marek-safar asked for a repro outside of the context of that ticket.

### Current Behavior

Leaks approximately 7MB per 24 hours.

### Expected Behavior

Memory usage remains static.

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

**Version Used**:

```
Mono JIT compiler version 5.20.1.34 (tarball Fri Aug  2 01:15:49 UTC 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       normal
        Notifications: epoll
        Architecture:  armel,vfp+hard
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          supported, not enabled.
        Suspend:       preemptive
        GC:            sgen (concurrent by default)
```

```
Mono JIT compiler version 5.20.1.19 (tarball Thu Apr 11 09:02:17 UTC 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          yes(600)
        Suspend:       hybrid
        GC:            sgen (concurrent by default)
```

```
Mono JIT compiler version 6.0.0.313 (tarball Sun Jul 14 09:57:59 UTC 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          yes(600)
        Suspend:       hybrid
        GC:            sgen (concurrent by default)
```

## Stacktrace

N/A

**Comment by filipnavara** [comment]
How is this a leak?! You keep producing tasks at faster pace than they can finish. Of course they are piling up and consuming memory.

**Comment by ChristianHackAMS** [comment]
No it isn't creating them faster than they can finish - you'll never have more than two running at once in fact with these numbers. The tasks run in parallel and still finish on time.

If each task was dependent on the previous one finishing, **then** you'd have an out of control task creation.

**Comment by filipnavara** [comment]
There is a limit on the parallelism. `Task.Run` says you want to run it but it does not guarantee that it will run this particular instant. The 50ms sleep block from the point when the task is scheduled to run, not from the point when the task is added to the queue.

**Comment by ShonkyCH** [comment]
All that really does is change what the 50ms delay effectively becomes. The task still finishes independently. There's still no out of control task creation. 

Try the example code - on ARM linux it leaks memory.

**Comment by filipnavara** [comment]
I get what you are trying to say... if you have at least two parallel threads processing the tasks and only one producing them then you should not reach the point of resource starvation (assuming a balanced scheduler).

**Comment by omfgicbf** [comment]
Even if there were delays or issues caused by timer accuracy, or even if the 50ms was increased to 100ms, all that will do is increase the number of threads involved. The tasks themselves are interleaved, so provided we're not exceeding the thread pool allocation, or maxing out the capabilities of the processor, then this should run fine.

Simply watching the ThreadPool worker thread count whilst this is running demonstrates this point.

Adding
```
ThreadPool.GetAvailableThreads(out var wt, out var cpt);
ThreadPool.GetMaxThreads(out var maxwt, out var maxcpt);
ThreadPool.GetMinThreads(out var minwt, out var mincpt);

Console.WriteLine($"wt avail/min/max {wt,4}/{minwt,4}/{maxwt,4} cpt avail/min/max {cpt,4}/{mincpt,4}/{maxcpt,4}");
```

Gives this output:-
```
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max  100/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   98/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   98/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
wt avail/min/max   99/   1/ 100 cpt avail/min/max  100/   1/ 100
```

Further to this, running with a 10ms delay inside the Task, and a 40ms delay in the loop, does the same thing; this does not overlap, as the Task should have more than enough time to complete before the next is queued.

It also does *not* leak when run on Windows/.NET, but it does when run on Windows with Mono.

One last thing @filipnavara, I'd appreciate you reading https://github.com/mono/mono/blob/master/code_of_conduct.md before responding.

@marek-safar Why was this closed?

**Comment by filipnavara** [comment]
I apologize if I offended you. I didn't mean to do that.

People open here issues all the time with something that they claim to be memory leak when they actually just leak the memory themselves with keeping [hidden] references to objects allocated in a loop.

I simply stated what I thought was wrong with the example. You proved me that my reading of the code is wrong and that in fact the explanation with resource starvation is not explaining what you are observing.

**Comment by omfgicbf** [comment]
@filipnavara all good; thanks for following up.

**Comment by omfgicbf** [comment]
@filipnavara any idea on how to get this re-opened, or do I just log another one? @marek-safar seems to be ignoring it.

**Comment by omfgicbf** [comment]
Windows/.NET, ~19 hours runtime:-

![image](https://user-images.githubusercontent.com/39300379/62663446-d9c2f900-b9ba-11e9-8221-841dbb0289ba.png)

Linux/Mono 5.20.1.34 ARM, 10 hours runtime:-

![image](https://user-images.githubusercontent.com/39300379/62663513-1e4e9480-b9bb-11e9-9d45-282db67755d0.png)

Linux/Mono 6.0.0.313 x86, ~19 hours runtime:-

![image](https://user-images.githubusercontent.com/39300379/62663546-48a05200-b9bb-11e9-8de4-ba1ed7a3e89d.png)

**Comment by MattL0** [comment]
did you try 3.0.0.319?

**Comment by omfgicbf** [comment]
@MattL0 not yet. Is there a changelog for 6.0.0.313 => 6.0.0.319 that suggests any fixes have been added that might be related to this?

**Comment by MattL0** [comment]
not sure but : 

https://github.com/mono/mono/issues/15751

**Comment by omfgicbf** [comment]
Thanks @MattL0, I checked out the issue you referenced, and it's probably unrelated. We've had this issue from at least 5.18.

Nonetheless, I'll run a test with 6.0.0.319 to see if it makes a difference.

**Comment by MattL0** [comment]
yes there was a memory leak in mono 5 too ( but less intense ). Hope you are on something here !

**Comment by omfgicbf** [comment]
OK, So using 6.0.0.319:

```
Mono JIT compiler version 6.0.0.319 (tarball Tue Aug  6 16:30:29 UTC 2019)
Copyright (C) 2002-2014 Novell, Inc, Xamarin Inc and Contributors. www.mono-project.com
        TLS:           __thread
        SIGSEGV:       altstack
        Notifications: epoll
        Architecture:  amd64
        Disabled:      none
        Misc:          softdebug
        Interpreter:   yes
        LLVM:          yes(600)
        Suspend:       hybrid
        GC:            sgen (concurrent by default)
```

I get a slightly different result (below over ~20 hours):-

![image](https://user-images.githubusercontent.com/39300379/62742180-e90c7a00-ba80-11e9-9b23-aed0c0ef22b6.png)

This is similar to the result I got with 6.0.0.313, however, that test didn't run as long as this. So perhaps I need to run them longer to see the drop-off.

Nonetheless, the earlier versions (5.x) don't have the drop at all (as we're seeing in our own program, which shows continued growth over weeks).

Where it gets even weirder, is when I run the exact same program (as represented in the graph above) on the same system, yet with profiling turned on (`mono --profile=log:heapshot=300000ms,alloc test.exe`), it does this:-

![image](https://user-images.githubusercontent.com/39300379/62742345-9089ac80-ba81-11e9-836c-3cb6e7370003.png)

**Comment by n432sean** [comment]
The drop offs suggest to me that the memory is eventually garbage collected in a higher generation than would be preferred. The memory is probably promoted to a higher generation due to the Thread.Sleeps. The problem may not occur with heapshot, because heapshot may be running the garbage collector more frequently.

**Comment by omfgicbf** [comment]
That was the conclusion I came to as well, however, all that suggests is that GC is broken (or "not tuned for this workload" as others will undoubtedly put it).

I'm not going to stand by my comment because I don't have the energy to argue why this reproducible behaviour is undesirable and in our case unsupportable.

I think I may have even tried some manual GC to see if that fixed things, but from memory (no pun intended) it didn't help in the version we're running, and causes a bunch of other issues.

I gave up on this because every ticket one raises has either no response, a snarky response or "have you tried 6.x?"; there is zero interest in fixing bugs in 5.x and for (many) reasons I don't want to go into, this isn't always a solution for many people.

I found that our overall issue was in fact related to long-running tasks spawned by `Task.Run`; this is related to an acknowledged issue from 2018 (referred to in the OP) where there is a complete lack of interest in resolving the issue.

The solution I found was to just not do what causes the leak until we can complete our port to .NET Core.

**Comment by GrahamBriggs** [comment]
I agree with omfgicbf.  This is a bug in Mono when using Task.Run in mono version 6.12.0.90 running on the Raspberry Pi (3 or 4 - Raspbian Buster)

I have a program that uses Task.Run to launch a computation in the background. I have logging to tell me if the calc takes longer than the frequency I am calling it and I am positive it finishes in about 10 ms and I am calling it at 20 Hz (50 ms). 

The program can run on Windows for days never consumes more than about 5 MB of memory.  When I run the same program on Raspberry Pi, it consumes memory very quickly.  I ran the program under the profiler, and every 100 seconds, there are another 1550 or so more System.Threading.IThreadPoolWorkItem as well as 1550 additional System.Threading.ThreadPoolWorkQueue.QueueSegment items in the heap shot.

![Screen Shot 2020-10-28 at 6 26 07 PM](https://user-images.githubusercontent.com/5171343/97514542-3574fb80-194c-11eb-86ed-e88be42e008c.png)

I refactored the program to not use Task.Run. There are however many async and await calls remaining in the program. It now consumes less memory, but it is still has problems.


![Screen Shot 2020-10-28 at 7 03 38 PM](https://user-images.githubusercontent.com/5171343/97516524-7f5fe080-1950-11eb-8dfe-c2e99f00ae26.png)


I have not tested this specific program on Mono 5 yet, but for the last couple of years I have had another program that uses Task.Run in the exact same way. This program does not consume excess memory when running on Mono.

I respectfully suggest that Mono team take another look at this issue and evaluate it as a bug.

**Comment by cmcqueen** [comment]
What is the status of this issue? Has the Mono team tried reproducing and investigating this reported issue?

I am working on a project on an ARM embedded Linux platform which has exhibited a memory leak on Mono since 4.8.x onwards. We have been sticking with Mono 4.6.x which doesn't have the issue. Meanwhile, we've been trying to make a simple example-of-bug program, but that has been elusive. It seems to involve some combination of timers and work queues.

Now I've found this issue, and it might be related. I'm surprised it hasn't been investigated and resolved after all this time.

**Comment by mcrossley** [comment]
Any update on this, my long running program exhausts a rPi 4 memory in two weeks and either crashes or crashes the whole system. This leak occurs on Mono 5.x and 6.x.  There is no leak at all when running on Windows.
